File: mobile/src/api/queries/portfolio.ts
import { useQuery } from '@tanstack/react-query';
import client from '../client';
import { STALE_TIMES } from '../../utils/constants';

export interface Portfolio {
  id: string;
  name: string;
  initial_capital: number;
  current_cash: number;
  invested_value: number;
  current_value: number;
  total_pnl: number;
  total_pnl_pct: number;
  day_pnl: number;
  benchmark_symbol: string;
  total_trades: number;
  win_rate: number;
  is_active: boolean;
}

export interface Holding {
  id: string;
  instrument_id: string;
  quantity: number;
  avg_buy_price: number;
  invested_value: number;
  current_price: number | null;
  current_value: number | null;
  pnl: number | null;
  pnl_pct: number | null;
  day_change_pct: number | null;
  weight_pct: number | null;
}

export interface Snapshot {
  date: string;
  total_value: number | null;
  cash: number | null;
  day_pnl: number | null;
  day_pnl_pct: number | null;
  cumulative_pnl_pct: number | null;
  benchmark_value: number | null;
  benchmark_pnl_pct: number | null;
}

export function usePortfolio() {
  return useQuery<Portfolio>({
    queryKey: ['portfolio'],
    queryFn: () => client.get('/portfolio/').then((r) => r.data),
    staleTime: STALE_TIMES.portfolio,
  });
}

export function useHoldings() {
  return useQuery<Holding[]>({
    queryKey: ['holdings'],
    queryFn: () => client.get('/portfolio/holdings').then((r) => r.data),
    staleTime: STALE_TIMES.portfolio,
  });
}

export function usePortfolioHistory(days = 30) {
  return useQuery<Snapshot[]>({
    queryKey: ['portfolio-history', days],
    queryFn: () => client.get(`/portfolio/history?days=${days}`).then((r) => r.data),
    staleTime: 300_000,
  });
}

File: src/api/schemas/portfolio.py
"""Pydantic schemas for portfolio and holdings."""
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel


class HoldingResponse(BaseModel):
    id: uuid.UUID
    instrument_id: uuid.UUID
    quantity: int
    avg_buy_price: float
    invested_value: float
    current_price: float | None
    current_value: float | None
    pnl: float | None
    pnl_pct: float | None
    day_change_pct: float | None
    weight_pct: float | None

    class Config:
        from_attributes = True


class PortfolioResponse(BaseModel):
    id: uuid.UUID
    name: str
    initial_capital: float
    current_cash: float
    invested_value: float
    current_value: float
    total_pnl: float
    total_pnl_pct: float
    day_pnl: float
    benchmark_symbol: str
    total_trades: int
    win_rate: float
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class SnapshotResponse(BaseModel):
    date: date
    total_value: float | None
    cash: float | None
    invested_value: float | None
    day_pnl: float | None
    day_pnl_pct: float | None
    cumulative_pnl_pct: float | None
    benchmark_value: float | None
    benchmark_pnl_pct: float | None
    num_holdings: int | None

    class Config:
        from_attributes = True
